### PR TITLE
[DEMO] Added support: MXNet, ONNX, OpenCV, PyTorch, TensorFlow, TVM

### PR DESCRIPTION
**Pull request by Vladislav-Denisov**
# Изменения:
- Внесены общие изменения в логику запуска демо (для поддержки запуска различных фреймворков и переиспользования кода).
## Фреймворки:
- MXNet - добавлен в демо. Внесены правки в docker-файл для проверки работоспособности `accuracy_checker`.
- ONNX - добавлен в демо. Внесены правки в docker-файл для проверки работоспособности  `accuracy_checker`. А также непосредственно в обертку для запуска `accuracy_checker` в `dl-benchmark`.
- OpenCV - добавлен в демо. Внесены правки в docker-файл для проверки работоспособности  `accuracy_checker`.
- TensorFlow - добавлен в демо на примере запуска для модели фреймворка TF2.  Лаунчер для `accuracy_check` изменен.
- PyTorch - добавлен в демо. Внесены правки в docker-файл для проверки работоспособности  `accuracy_checker`.
- TVM - добавлен в демо. Внесены правки в docker-файл для поддержки `accuracy_checker` с кастомной реализацией.

### CODE REVIEW

File: demo/linux.sh [hunk 1]
```diff
@@ -1,18 +1,55 @@
...
+supported_frameworks="OpenVINO_DLDT TensorFlow MXNet ONNXRuntime OpenCV"
+
+usage() {
+    echo "Usage: $0 [-l LOGIN] [-p PASSWORD] [-f FRAMEWORK] [-d GIT_LINK_TO_DATASET]"
+    echo "Options:"
+    echo "  -l          Login of the current user."
+    echo "  -p          Password of the current user."
+    echo "  -f          Framework (supported: $supported_frameworks)."
+    echo "  -d          The address to the GitHub repository, which contains datasets for benchmarking."
+    echo "              It is required that the Databases/ImageNet/ directory be created"
```
Comment (valentina-kustikova) [inline demo/linux.sh:12]: @Vladislav-Denisov, у меня пришла мысль, может, нам стоит хранить пару изображений для бенчмаркинга прямо рядом с демо, чтобы не зависеть от других ресурсов, а для определения качества просто отдавать путь до папки с загруженными картинками. Что вы думаете?

Еще момент - тестовая модель работает с данными Cifar, а не с ImageNet. При этом везде используется набор ImageNet. Права?

File: demo/linux.sh [hunk 2]
```diff
@@ -1,18 +1,55 @@
...
+    echo "Framework '$framework' is not supported."
+    exit_abnormal
+fi
+
 
 echo "[ INFO ] Demo application has been started"
 demo_folder="$PWD"
-root_folder="${demo_folder}/../.."
+root_folder="${demo_folder}/.."
 openvino_version="2022.1.0"
```
Comment (valentina-kustikova) [inline demo/linux.sh:53]: Стоит переехать на более новую версию OpenVINO

File: src/benchmark/executors.py [hunk 3]
```diff
@@ -24,6 +24,10 @@ def get_executor(executor_type, log):
         raise ValueError('Executor type must be from list: host_machine, docker_container')
 
     def set_target_framework(self, target_framework):
+        if target_framework == 'ONNX Runtime Python' or target_framework == 'ONNX Runtime CPP':
+            target_framework = 'ONNXRuntime'
+        elif target_framework == 'OpenCV DNN Python' or target_framework == 'OpenCV DNN CPP':
+            target_framework = 'OpenCV'
```
Comment (valentina-kustikova) [inline src/benchmark/executors.py:30]: Вот это неправильно, потому что для C++ API библиотек ONNX Runtime и OpenCV запуск выполняется с использованием плюсовой реализации вывода.

Comment (Vladislav-Denisov) [reply src/benchmark/executors.py:30]: Сейчас в моей предложенной реализации происходит следующее:

1. В `/demo/benchmark_configs/ONNXRuntime.xml` указан `InferenceFramework="ONNX Runtime Python"` (можно добавить второй тест для `ONNX Runtime` - для `Cpp` реализации вывода).
2. В `/demo/linux.sh` в качестве параметра `-f` указывается фреймворк `ONNXRuntime` без явного указания реализации, так как она прописана в конфиге из п.1.
3. Для создания докер-контейнера в `/demo/linux.sh` в качестве имени контейнера применяется переданное значение параметра `ONNXRuntime`. Причем согласно `/docker/ONNXRuntime/Dockerfile` контейнер всегда содержит в себе сразу обе реализации: и `Python`, и `Cpp`.
4. Затем при процессе `/src/deployment/deploy.py` в том числе забирается конфиг из п.1 - то есть при запуске бенчмарка точно будет известна конкретная требуемая реализация вывода средствами `ONNXRuntime`: Python (или Cpp).
А в параметре `-n` скрипта `deploy.py` точно указывается наименование контейнера из п.3 - `ONNXRuntime` (без каких-либо доп уточнений).
5. Далее в `/demo/linux.sh` есть вызов скрипта `/src/remote_control/remote_start.py` c конфигом, включающим `<Executor>docker_container</Executor>`. Выполняется запуск экспериментов на "удаленных" машинах в соответствующих докер-контейнерах.
6. При запуске скрипта `/src/benchmark/inference_benchmark.py` создается требуемый `Executor` (`HostExecutor` или `DockerExecutor`).
`HostExecutor` никак не использует `target_framework`.
`DockerExecutor` по `target_framework` ищет нужный контейнер. Причем `target_framework` в этот момент содержит в себе значение из конфига, упомянутого в п.1. Если не выполнить предварительную замену, которую я добавил, то такой контейнер найти не получится.
Непосредственно на запуск вывода "правильной" реализацией (`Python` или `Cpp`) это не влияет - обращение к конфигу каждого теста выполняется чуть позже:
https://github.com/itlab-vision/dl-benchmark/blob/b0d60e65c902f6282a739560569b39c12a2d1873/src/benchmark/inference_benchmark.py#L70

В результате при таком подходе мы имеем один контейнер, содержащий в себе сразу обе реализации `ONNXRuntime`.
А также корректные конфиги и работу с ними при запуске бенчмарка. Т.е. если хотим использовать Python реализацию, значит просто указываем ее в конфиге и всё взлетит.
Аналогично для MXNet.

Поправьте меня, если я где-то ошибся.

Comment (valentina-kustikova) [reply src/benchmark/executors.py:30]: Дело не в контейнере и не в демо. Метод этот вызывается в файле src/benchmark/frameworks/processes.py, когда создается процесс для запуска конкретного теста бенчмаркинга. Если мы вставим ваше условие, то не понятно, какой Executor будет вызван при выполнении теста, потому что этот метод явно вызывается перед запуском теста, и в зависимости от этого параметра запускается либо banchmark_cpp, либо скрипт с соответствующей реализации вывода.

Comment (Vladislav-Denisov) [reply src/benchmark/executors.py:30]: Насколько я смог разобраться:

Точка входа `src/benchmark/inference_benchmark.py`
1.1 Создаем `process_executor` - в результате появится объект `HostExecutor` или `DockerExecutor` (о том, что target_framework используется только для поиска правильного докер-контейнера я писал в своём предыдущем сообщении в п.6)
https://github.com/itlab-vision/dl-benchmark/blob/f8994ef4988abdd9df7c29063cdd8bdf2c9c9f07/src/benchmark/inference_benchmark.py#L64
1.2 Затем согласно конфиг-файлу заполняется `framework_name`. Причем в конфиге указывается полное название, с учетом конкретной реализации, например, `ONNX Runtime Python`.
https://github.com/itlab-vision/dl-benchmark/blob/f8994ef4988abdd9df7c29063cdd8bdf2c9c9f07/src/benchmark/inference_benchmark.py#L70
1.2 Далее по только что заполненному и никак не измененному `framework_name` создается объект процесса `test_process`, который будет исполнять "нужную специализацию инференса". 
https://github.com/itlab-vision/dl-benchmark/blob/f8994ef4988abdd9df7c29063cdd8bdf2c9c9f07/src/benchmark/inference_benchmark.py#L77
Причем `process_executor`, упомянутый в п.1.1 передается как параметр.
Например, для:
-- `FrameworkWrapperRegistry()['ONNX Runtime Python']` будет создан объект класса `ONNXRuntimePythonWrapper` и согласно `/src/benchmark/frameworks/onnx_runtime_python/onnx_runtime_python_process.py` этот объект в дальнейшем будет использовать скрипт `inference_onnx_runtime.py`
https://github.com/itlab-vision/dl-benchmark/blob/f8994ef4988abdd9df7c29063cdd8bdf2c9c9f07/src/benchmark/frameworks/onnx_runtime_python/onnx_runtime_python_process.py#L21
--  `FrameworkWrapperRegistry()['ONNX Runtime']` будет создан объект класса `OnnxRuntimeProcess` и согласно `/src/benchmark/frameworks/onnx_runtime/onnx_runtime_process.py` будет использовать скрипт для cpp-реализации.
https://github.com/itlab-vision/dl-benchmark/blob/f8994ef4988abdd9df7c29063cdd8bdf2c9c9f07/src/benchmark/frameworks/onnx_runtime/onnx_runtime_process.py#L25
https://github.com/itlab-vision/dl-benchmark/blob/f8994ef4988abdd9df7c29063cdd8bdf2c9c9f07/src/benchmark/frameworks/onnx_runtime/onnx_runtime_process.py#L45
1.3 Затем для этого созданного объекта `test_process` вызывается метод `execute`
https://github.com/itlab-vision/dl-benchmark/blob/f8994ef4988abdd9df7c29063cdd8bdf2c9c9f07/src/benchmark/inference_benchmark.py#L81
И согласно `src/benchmark/frameworks/processes.py`
https://github.com/itlab-vision/dl-benchmark/blob/f8994ef4988abdd9df7c29063cdd8bdf2c9c9f07/src/benchmark/frameworks/processes.py#L56
будет сформирована правильная command-line для запуска инференса. Ведь `_fill_command_line` будет вызван от соответствующего дочернего класса `ProcessHandler` - от  `OnnxRuntimeProcess` или от  `ONNXRuntimePythonWrapper`.
В результате получается, что
https://github.com/itlab-vision/dl-benchmark/blob/f8994ef4988abdd9df7c29063cdd8bdf2c9c9f07/src/benchmark/frameworks/processes.py#L63
действительно влияет только на определение контейнера в случае `process_executor=DockerExecutor`

File: demo/linux.sh (hunk 2 above)
Comment (Vladislav-Denisov) [reply demo/linux.sh]: Версию обновил

File: demo/linux.sh (hunk 1 above)
Comment (Vladislav-Denisov) [reply demo/linux.sh:12]: Предлагаю пока не менять этот подход. Насколько я понимаю, ключ `-d` был предусмотрен для корректного построения docker-контейнеров.
https://github.com/itlab-vision/dl-benchmark/blob/f8994ef4988abdd9df7c29063cdd8bdf2c9c9f07/docker/Dockerfile#L70-L71

Бенчмарк запускается на изображениях из нашего приватного репозитория (ImageNet), валидация на Cifar (потому что пример для AccuracyChecker в репозитории OMZ использует именно этот набор данных).

File: src/benchmark/executors.py (hunk 3 above)
Comment (Vladislav-Denisov) [reply src/benchmark/executors.py]: Откатился к существующей схеме работы `src/benchmark/executors.py` в https://github.com/itlab-vision/dl-benchmark/pull/552/commits/32bcd3784cbf4ed8ae97f4c427a721fc414dd880

File: demo/accuracy_checker_configs/ONNX_Runtime_Python.xml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    <Test>
+        <Model>
+            <Task>classification</Task>
+            <Name>SampleNet</Name>
+            <Precision>FP32</Precision>
+            <SourceFramework>ONNXRuntime</SourceFramework>
+            <Directory>/media/models</Directory>
+        </Model>
+        <Parameters>
+            <InferenceFramework>ONNX Runtime Python</InferenceFramework>
```
Comment (valentina-kustikova) [inline demo/accuracy_checker_configs/ONNX_Runtime_Python.xml:12]: В Accuracy Checker мы не разделяем Python и CPP APIs, мы просто валидируем корректность работы с использованием какого-то программного интерфейса библиотеки.

File: src/accuracy_checker/process.py [hunk 5]
```diff
@@ -9,9 +9,12 @@ def __init__(self, log, executor, test):
         self._test = test
         self._output = None
         self._status = None
-        self.supported_frameworks = {'OpenVINO DLDT': 'dlsdk', 'Caffe': 'caffe', 'TensorFlow': 'tf',
-                                     'TensorFlow_Lite': 'tf_lite', 'MXNet': 'mxnet', 'OpenCV': 'opencv',
-                                     'PyTorch': 'pytorch', 'TVM': 'tvm'}
+        self.supported_frameworks = {'OpenVINO DLDT': 'dlsdk', 'Caffe': 'caffe', 'TensorFlow': 'tf2',
+                                     'TensorFlowLite': 'tf_lite', 'TensorFlowLite Cpp': 'tf_lite',
```
Comment (valentina-kustikova) [inline src/accuracy_checker/process.py:13]: В Accuracy Checker мы не разделяем Python и CPP APIs, мы просто валидируем корректность работы с использованием какого-то программного интерфейса библиотеки, который имеется в OpenVINO Accuracy Checker. Поэтому не имеет смысла порождать столько значений параметра. Формат модели не зависит от того API, который далее используется.

File: demo/accuracy_checker_configs/ONNX_Runtime_Python.xml (hunk 4 above)
Comment (Vladislav-Denisov) [reply demo/accuracy_checker_configs/ONNX_Runtime_Python.xml:12]: Обсудили на прошлом созвоне. Если не разделить в явном виде, то при запуске в докер-контейнере скрипт не сможет определить его наименование.

File: demo/benchmark_configs/ONNX_Runtime_Python.xml [hunk 6]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Tests>
+    <Test>
+        <Model>
+            <Task>classification</Task>
+            <Name>SampleNet</Name>
+            <Precision>FP32</Precision>
+            <SourceFramework>ONNXRuntime</SourceFramework>
```
[the remark this answers is not in the document]
Comment (Vladislav-Denisov) [reply demo/benchmark_configs/ONNX_Runtime_Python.xml:8]: Не знаю насколько это критично в рамках конфига для демо.
В офф репозитории https://github.com/openvinotoolkit/open_model_zoo/tree/master/tools/accuracy_checker/data/test_models для этих SampleNet'ов нет никакой README-шки.

**Comment by Vladislav-Denisov** [comment]
@valentina-kustikova на данный момент в этом PR есть всё необходимое для демо с использованием фреймоврков, перечисленных в первом сообщении.
Исключение - TensorFlowLite - его надо посмотреть отдельно.
Думаю, можно мерджить.

**Review by valentina-kustikova — DISMISSED** [review]

**Comment by valentina-kustikova** [comment]
@ismukhin, посмотрите, пожалуйста.

File: docker/TVM/Dockerfile [hunk 7]
```diff
@@ -72,5 +72,12 @@ RUN if [ "$CAFFE" = "true"  ]; then \
     /bin/bash -c "conda install -n tvm-env -y -c anaconda libgcc-ng=11.2.0"; \
     /bin/bash -c "conda install -n tvm-env -y -c defaults protobuf=3.17.2 libprotobuf=3.17.2 requests=2.27.1"; \
     fi
-    
+
+WORKDIR /tmp/open_model_zoo/
+RUN git remote add omz_custom_tvm https://github.com/a-sidorova/open_model_zoo.git && \
```
Comment (ismukhin) [inline docker/TVM/Dockerfile:77]: Мне кажется, ветку с ```accuracy_checker'ом``` для ```tvm``` нужно форкнуть в itlab-vision, чтобы в будущем не оказалось, что его больше нет :) (думаю, сделать и держать его у нас имеет смысл до момента включения ```tvm``` в мастер ```open_model_zoo```)

Comment (a-sidorova) [reply docker/TVM/Dockerfile:77]: Создала бранчу в `open_model_zoo_tvm` :grinning: (коммиты поверх 2024.3 релиза).
https://github.com/itlab-vision/open_model_zoo_tvm/tree/24.3.0/tvm

Comment (Vladislav-Denisov) [reply docker/TVM/Dockerfile]: @ismukhin заменил ссылку на репозиторий в https://github.com/itlab-vision/dl-benchmark/pull/552/commits/cfc546985931283a6d7f174423621965e47dfad2

**Review by valentina-kustikova — APPROVED** [review]

**Comment by valentina-kustikova** [comment]
@ismukhin, посмотрите, если все ок, то замержите, пожалуйста.
